File: src/worker.py
"""Load one unit (one <layer>_<type>) into its table.

For each record (from the wrapper's repeated field), flatten the generic header
sub-message and the payload sub-message into one flat row, prefixed with run_id
and a parsed ts, suffixed with the ingest time. Column order matches
TableSchema.columns.

Usable in-process (tests, single worker) or in a child process (parallel) — both
call `load_unit`, so the decode/row-building path is identical and tested once.
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path

from .errors import InputError
from .reader import iter_records
from .registry import TableSchema
from .store.base import Store

# Header timestamp. Producers have shipped a few shapes, so we try several:
#   "Mon Jul 27 18:02:21 2026"        (abbreviated names + year)
#   "Monday July 27 18:01:46:233"     (full names, NO year, trailing ms)
#   ISO-8601                          "2026-07-27T18:02:21"
# (fmt, has_year); no-year formats get the current year injected.
_TS_FORMATS = [
    ("%a %b %d %H:%M:%S %Y", True),
    ("%A %B %d %H:%M:%S %Y", True),
    ("%Y-%m-%dT%H:%M:%S", True),
    ("%Y-%m-%d %H:%M:%S", True),
    ("%a %b %d %H:%M:%S", False),
    ("%A %B %d %H:%M:%S", False),
]
_BRACKETS = re.compile(r"^\[(.*)\]$")
# a millisecond group tacked onto the time, e.g. "18:01:46:233" or "18:01:46.233"
_MILLIS = re.compile(r"(\d{2}:\d{2}:\d{2})[:.]\d{1,6}\b")
# a stray offset appended after 'Z' (e.g. "...Z00:00" or "...Z05:30"). RFC 3339
# uses either 'Z' (UTC) OR an offset, never both — this is a producer bug. 'Z'
# already means UTC, so we drop the trailing offset and keep 'Z'.
_Z_OFFSET = re.compile(r"Z[+-]?\d{2}:?\d{2}$")

# ClickHouse DateTime is an unsigned 32-bit epoch: 1970-01-01 .. 2106-02-07 UTC.
# We treat the (tz-less) header string as UTC so the stored ts is deterministic
# regardless of the loader machine's timezone, and clamp to this range so a bad
# timestamp can never serialize out of bounds (which crashes the insert).
_UTC = timezone.utc
_DT_MIN = datetime(1970, 1, 1, tzinfo=_UTC)
_DT_MAX = datetime(2106, 2, 7, tzinfo=_UTC)


@dataclass
class UnitResult:
    stem: str
    files: int
    records: int                    # parent records read
    table_rows: dict                # table name -> rows inserted (main + children)

    @property
    def ok(self) -> bool:
        # reconcile: the main table got exactly one row per parent record
        return self.table_rows.get(self.stem, 0) == self.records

    @property
    def total_rows(self) -> int:
        return sum(self.table_rows.values())


def parse_ts(s: str) -> datetime | None:
    """Parse the header timestamp string to a datetime (aware if the string
    carries a timezone), or None if it matches no known format. The original
    string is always kept in the `timestamp` column, so a parse miss loses
    nothing."""
    if not s:
        return None
    s = s.strip()
    m = _BRACKETS.match(s)
    if m:
        s = m.group(1).strip()
    s = _Z_OFFSET.sub("Z", s)   # normalize malformed "...Z00:00" -> "...Z"

    # ISO 8601 / RFC 3339 first — handles fractional seconds and a Z/offset
    # timezone (e.g. 2026-07-31T16:33:53.005Z or ...+05:30). The Z rewrite keeps
    # this working on Python < 3.11 too.
    iso = (s[:-1] + "+00:00") if s.endswith("Z") else s
    try:
        return datetime.fromisoformat(iso)
    except ValueError:
        pass

    # Fallback: human-readable formats. strptime here has no fractional-seconds
    # slot, so drop a trailing millisecond group first.
    s2 = _MILLIS.sub(r"\1", s)
    for fmt, has_year in _TS_FORMATS:
        try:
            dt = datetime.strptime(s2, fmt)
        except ValueError:
            continue
        if not has_year:
            dt = dt.replace(year=datetime.now().year)
        return dt
    return None


def ts_for_db(s: str) -> datetime:
    """The `ts` column value: a UTC-aware DateTime, clamped to ClickHouse's
    representable range so it always serializes to a valid unsigned epoch.
    Unparseable timestamps become the epoch (1970-01-01), never a crash."""
    dt = parse_ts(s)
    if dt is None:
        return _DT_MIN
    # A tz-aware timestamp (ISO with Z/offset) is converted to the real UTC
    # instant; a naive one is interpreted as UTC. Either way ts is deterministic.
    dt = dt.astimezone(_UTC) if dt.tzinfo is not None else dt.replace(tzinfo=_UTC)
    if dt < _DT_MIN:
        return _DT_MIN
    if dt > _DT_MAX:
        return _DT_MAX
    return dt


def load_unit(pb_parts: list[Path], schema: TableSchema, store: Store,
              run_id: str, batch_size: int) -> UnitResult:
    """Load one unit into its main table plus a child table per repeated payload
    sub-message. A `record_id` (UInt64 sequence, per record) links a main row to
    its child rows."""
    store.ensure_table(schema.stem, schema.columns, schema.order_by)
    for ch in schema.children:
        store.ensure_table(ch.table, ch.columns, ch.order_by)

    loaded_at = datetime.now().replace(microsecond=0)
    generic_field = schema.generic_field
    payload_field = schema.payload_field
    generic_fields = schema.generic_fields
    payload_fields = schema.payload_fields
    has_rid = schema.has_record_id
    children = schema.children

    main_cols = [c.name for c in schema.columns]
    child_cols = {ch.table: [c.name for c in ch.columns] for ch in children}

    table_rows: dict[str, int] = {schema.stem: 0}
    for ch in children:
        table_rows[ch.table] = 0
    main_batch: list[tuple] = []
    child_batches: dict[str, list[tuple]] = {ch.table: [] for ch in children}

    def flush_main():
        if main_batch:
            store.insert(schema.stem, main_cols, main_batch)
            table_rows[schema.stem] += len(main_batch)
            main_batch.clear()

    def flush_child(ch: "object"):
        b = child_batches[ch.table]
        if b:
            store.insert(ch.table, child_cols[ch.table], b)
            table_rows[ch.table] += len(b)
            b.clear()

    records = 0
    record_id = 0
    for part in pb_parts:
        for rec in iter_records(part, schema):
            records += 1
            g = getattr(rec, generic_field)
            p = getattr(rec, payload_field)
            ts = ts_for_db(g.timestamp)
            gen_vals = tuple(getattr(g, f) for f in generic_fields)

            if has_rid:
                main_batch.append((run_id, ts, record_id, *gen_vals,
                                   *(getattr(p, f) for f in payload_fields), loaded_at))
            else:
                main_batch.append((run_id, ts, *gen_vals,
                                   *(getattr(p, f) for f in payload_fields), loaded_at))

            for ch in children:
                cb = child_batches[ch.table]
                for sub in getattr(p, ch.repeated_field):
                    cb.append((run_id, ts, record_id, *gen_vals,
                               *(getattr(sub, f) for f in ch.sub_fields), loaded_at))
                if len(cb) >= batch_size:
                    flush_child(ch)

            record_id += 1
            if len(main_batch) >= batch_size:
                flush_main()

    flush_main()
    for ch in children:
        flush_child(ch)

    return UnitResult(stem=schema.stem, files=len(pb_parts),
                      records=records, table_rows=table_rows)

File: internal/worker/worker.go
// Package worker loads one unit (one <layer>_<type>) into its table.
//
// For each record (from the wrapper's repeated field), the generic header
// sub-message and the payload sub-message are flattened into one flat row,
// prefixed with run_id and a parsed ts, suffixed with the ingest time. Column
// order matches TableSchema.Columns.
package worker

import (
	"context"
	"regexp"
	"strings"
	"time"

	"google.golang.org/protobuf/reflect/protoreflect"

	"github.com/vishal9359/Analysis-FW/internal/reader"
	"github.com/vishal9359/Analysis-FW/internal/registry"
	"github.com/vishal9359/Analysis-FW/internal/store"
)

// Header timestamp. Producers have shipped a few shapes, so we try several:
//
//	"Mon Jul 27 18:02:21 2026"        (abbreviated names + year)
//	"Monday July 27 18:01:46:233"     (full names, NO year, trailing ms)
//	ISO-8601 / RFC 3339               "2026-07-27T18:02:21.000Z"
//
// Layouts without a year get the current year injected.
var tsLayouts = []struct {
	layout  string
	hasYear bool
}{
	{"2006-01-02T15:04:05.999999999Z07:00", true},
	{"2006-01-02T15:04:05Z07:00", true},
	{"2006-01-02T15:04:05", true},
	{"2006-01-02 15:04:05", true},
	{"Mon Jan 2 15:04:05 2006", true},
	{"Monday January 2 15:04:05 2006", true},
	{"Mon Jan 2 15:04:05", false},
	{"Monday January 2 15:04:05", false},
}

var (
	bracketed = regexp.MustCompile(`^\[(.*)\]$`)
	// a millisecond group tacked onto the time, e.g. "18:01:46:233"
	millis = regexp.MustCompile(`(\d{2}:\d{2}:\d{2})[:.]\d{1,6}\b`)
	// a stray offset appended after 'Z' (e.g. "...Z00:00"). RFC 3339 uses
	// either 'Z' (UTC) OR an offset, never both — this is a producer bug. 'Z'
	// already means UTC, so drop the trailing offset and keep 'Z'.
	zOffset = regexp.MustCompile(`Z[+-]?\d{2}:?\d{2}$`)
	// collapse repeated spaces so "Jul  2" matches a single-space layout
	spaces = regexp.MustCompile(`\s+`)
)

// ParseTS parses a header timestamp string, or reports ok=false if it matches
// no known format. The original string is always kept in the `timestamp`
// column, so a parse miss loses nothing.
func ParseTS(s string) (time.Time, bool) {
	s = strings.TrimSpace(s)
	if s == "" {
		return time.Time{}, false
	}
	if m := bracketed.FindStringSubmatch(s); m != nil {
		s = strings.TrimSpace(m[1])
	}
	s = zOffset.ReplaceAllString(s, "Z") // normalize malformed "...Z00:00"

	for _, l := range tsLayouts {
		candidate := s
		if !strings.Contains(l.layout, ".999") {
			// layouts without a fractional slot: drop a trailing ms group
			candidate = millis.ReplaceAllString(candidate, "$1")
		}
		candidate = spaces.ReplaceAllString(candidate, " ")
		t, err := time.Parse(l.layout, candidate)
		if err != nil {
			continue
		}
		if !l.hasYear {
			t = t.AddDate(time.Now().Year(), 0, 0)
		}
		return t, true
	}
	return time.Time{}, false
}

// TSForDB is the `ts` column value: a UTC time clamped to tsRange (the target
// store's representable range) so it always serializes successfully.
// Unparseable timestamps become the range's lower bound, never a failure.
func TSForDB(s string, tsRange [2]time.Time) time.Time {
	lo, hi := tsRange[0], tsRange[1]
	t, ok := ParseTS(s)
	if !ok {
		return lo
	}
	// A tz-aware timestamp is converted to the real UTC instant; a naive one is
	// interpreted as UTC. Either way ts is deterministic regardless of the
	// loader machine's timezone.
	t = t.UTC()
	if t.Before(lo) {
		return lo
	}
	if t.After(hi) {
		return hi
	}
	return t
}

// UnitResult is what one unit's load produced.
type UnitResult struct {
	Stem      string
	Files     int
	Records   int            // parent records read
	TableRows map[string]int // table name -> rows inserted (main + children)
}

// OK reports whether the main table got exactly one row per parent record.
func (r UnitResult) OK() bool { return r.TableRows[r.Stem] == r.Records }

// TotalRows is every row written across main and child tables.
func (r UnitResult) TotalRows() int {
	n := 0
	for _, v := range r.TableRows {
		n += v
	}
	return n
}

// LoadUnit loads one unit into its main table plus a child table per repeated
// payload sub-message. A record_id (64-bit sequence, per record) links a main
// row to its child rows.
func LoadUnit(ctx context.Context, pbParts []string, schema *registry.TableSchema,
	st store.Store, runID string, batchSize int) (UnitResult, error) {

	res := UnitResult{Stem: schema.Stem, Files: len(pbParts), TableRows: map[string]int{schema.Stem: 0}}

	if err := st.EnsureTable(ctx, schema.Stem, schema.Columns, schema.OrderBy); err != nil {
		return res, err
	}
	for _, c := range schema.Children {
		if err := st.EnsureTable(ctx, c.Table, c.Columns, c.OrderBy); err != nil {
			return res, err
		}
		res.TableRows[c.Table] = 0
	}

	loadedAt := time.Now().Truncate(time.Second)
	tsRange := st.TimestampRange() // the target store's limits, not ours

	mainCols := columnNames(schema.Columns)
	childCols := map[string][]string{}
	for _, c := range schema.Children {
		childCols[c.Table] = columnNames(c.Columns)
	}

	mainBatch := make([][]interface{}, 0, batchSize)
	childBatch := map[string][][]interface{}{}
	for _, c := range schema.Children {
		childBatch[c.Table] = make([][]interface{}, 0, batchSize)
	}

	flushMain := func() error {
		if len(mainBatch) == 0 {
			return nil
		}
		if err := st.Insert(ctx, schema.Stem, mainCols, mainBatch); err != nil {
			return err
		}
		res.TableRows[schema.Stem] += len(mainBatch)
		mainBatch = mainBatch[:0]
		return nil
	}
	flushChild := func(table string) error {
		b := childBatch[table]
		if len(b) == 0 {
			return nil
		}
		if err := st.Insert(ctx, table, childCols[table], b); err != nil {
			return err
		}
		res.TableRows[table] += len(b)
		childBatch[table] = b[:0]
		return nil
	}

	// Reused across records; its values are copied into each row.
	genVals := make([]interface{}, len(schema.GenericFields))

	recordID := uint64(0)
	for _, part := range pbParts {
		records, err := reader.Records(part, schema)
		if err != nil {
			return res, err
		}
		for i := 0; i < records.Len(); i++ {
			rec := records.Get(i).Message()
			res.Records++

			g := rec.Get(schema.GenericField).Message()
			p := rec.Get(schema.PayloadField).Message()
			ts := TSForDB(g.Get(schema.TimestampFd).String(), tsRange)

			for j, path := range schema.GenericFields {
				genVals[j] = goValueAt(g, path)
			}

			row := make([]interface{}, 0, len(schema.Columns))
			row = append(row, runID, ts)
			if schema.HasRecordID {
				row = append(row, recordID)
			}
			row = append(row, genVals...)
			for _, path := range schema.PayloadFields {
				row = append(row, goValueAt(p, path))
			}
			row = append(row, loadedAt)
			mainBatch = append(mainBatch, row)

			for _, c := range schema.Children {
				subs := p.Get(c.RepeatedField).List()
				for k := 0; k < subs.Len(); k++ {
					sub := subs.Get(k).Message()
					crow := make([]interface{}, 0, len(c.Columns))
					crow = append(crow, runID, ts, recordID)
					crow = append(crow, genVals...)
					for _, path := range c.SubFields {
						crow = append(crow, goValueAt(sub, path))
					}
					crow = append(crow, loadedAt)
					childBatch[c.Table] = append(childBatch[c.Table], crow)
				}
				if len(childBatch[c.Table]) >= batchSize {
					if err := flushChild(c.Table); err != nil {
						return res, err
					}
				}
			}

			recordID++
			if len(mainBatch) >= batchSize {
				if err := flushMain(); err != nil {
					return res, err
				}
			}
		}
	}

	if err := flushMain(); err != nil {
		return res, err
	}
	for _, c := range schema.Children {
		if err := flushChild(c.Table); err != nil {
			return res, err
		}
	}
	return res, nil
}

func columnNames(cols []store.Column) []string {
	out := make([]string, len(cols))
	for i, c := range cols {
		out[i] = c.Name
	}
	return out
}

// goValueAt reads the scalar a FieldPath points at, descending through singular
// sub-messages (a 1:1 group flattened onto this row). An absent sub-message
// yields its fields' proto3 defaults rather than nil, matching a missing scalar.
func goValueAt(m protoreflect.Message, path registry.FieldPath) interface{} {
	for i := 0; i < len(path)-1; i++ {
		// Get on an unset message field returns a read-only empty message, so
		// the remaining reads produce zero values.
		m = m.Get(path[i]).Message()
	}
	return goValue(m, path[len(path)-1])
}

// goValue converts a protobuf field value to the concrete Go type the store
// driver expects for that column. A missing proto3 field yields its zero value,
// never nil — matching the Python loader's behaviour.
func goValue(m protoreflect.Message, fd protoreflect.FieldDescriptor) interface{} {
	v := m.Get(fd)
	switch fd.Kind() {
	case protoreflect.StringKind:
		return v.String()
	case protoreflect.BytesKind:
		return string(v.Bytes())
	case protoreflect.BoolKind:
		return v.Bool()
	case protoreflect.Int32Kind, protoreflect.Sint32Kind, protoreflect.Sfixed32Kind:
		return int32(v.Int())
	case protoreflect.Int64Kind, protoreflect.Sint64Kind, protoreflect.Sfixed64Kind:
		return v.Int()
	case protoreflect.Uint32Kind, protoreflect.Fixed32Kind:
		return uint32(v.Uint())
	case protoreflect.Uint64Kind, protoreflect.Fixed64Kind:
		return v.Uint()
	case protoreflect.FloatKind:
		return float32(v.Float())
	case protoreflect.DoubleKind:
		return v.Float()
	case protoreflect.EnumKind:
		return int32(v.Enum())
	default:
		return v.Interface()
	}
}
